Sources/MongoSwift/BSON/BSONDecoder.swift

```diff
@@ -3,17 +3,74 @@ import Foundation
 /// `BSONDecoder` facilitates the decoding of BSON into semantic `Decodable` types.
 public class BSONDecoder {
 
+    @available(OSX 10.12, *)
+    internal static var iso8601Formatter: ISO8601DateFormatter = {
+        let formatter = ISO8601DateFormatter()
+        formatter.formatOptions = .withInternetDateTime
+        return formatter
+    }()
+
+    /// Enum representing the different options for decoding `Date`s from BSON
+    public enum DateDecodingStrategy {
+        /// Decoding `Date`s stored as 64 bit integers counting the number of milliseconds since January 1, 1970
+        case millisecondsSince1970Int64
+
+        /// Decoding `Date`s stored as 64 bit integers counting the number of seconds since January 1, 1970
+        case secondsSince1970Int64
+
+        /// Decoding `Date`s stored as `Double`s counting the number of seconds since January 1, 1970
+        case millisecondsSince1970
+
+        /// Decoding `Date`s stored as `Double`s counting the number of milliseconds since January 1, 1970
+        case secondsSince1970
+
+        /// Decoding `Date`s by deferring to their default decoding implementation
+        case deferredToDate
+
+        /// Decoding `Date`s represented by ISO8601 formatted strings
+        @available(OSX 10.12, *)
+        case iso8601
+
+        /// Decoding `Date`s stored as strings parsable by the given formatter
+        case formatted(DateFormatter)
+
+        /// Decoding `Date`s using the provided closure
+        case custom((_ decoder: Decoder) throws -> Date)
+    }
+
+    /// Enum representing the different options for decoding `UUID`s from BSON
+    public enum UUIDDecodingStrategy {
+        /// Decoding `UUID`s by deferring to their default decoding implementation
+        case deferredToUUID
+
+        /// Decoding `UUID`s from strings
+        case fromString
+
+        /// Decoding `UUID`s from the BSON `Binary` type
+        case fromBinary
+    }
+
     /// Contextual user-provided information for use during decoding.
     public var userInfo: [CodingUserInfoKey: Any] = [:]
 
+    /// The strategy used for decoding dates with this instance.
+    public var dateDecodingStrategy: DateDecodingStrategy = .millisecondsSince1970Int64
```

Comment:
hmmm, good point - maybe we should specifically look for that case and give them a helpful error telling them to switch strategies? 
it is tempting to just make all of the other options additive (i.e. `secondsSince1970` actually means both `bsonDateTime` *and* `secondsSince1970`), but from an API perspective I think that might be more confusing... what do you think?

Reply:
I think it could be useful, but it's probably safer to keep a consistent API. In general, users aren't going to be setting a strategy for `Date`s unless they _know_ their data is stored in a certain nonstandard manner. Even if the `Date`s could be encoded in multiple ways, they always have `.custom` as a fallback.

Also, I don't think we need a custom error, as the current one is pretty decent, though it does say `Double` when really any number would work:
```
Fatal error: Error raised at top level: Swift.DecodingError.typeMismatch(Swift.Double, Swift.DecodingError.Context(codingPath: [CodingKeys
(stringValue: "b", intValue: nil)], debugDescription: "Expected to find a value that can be represented as a Double, but found value 2018-
12-20 17:20:25 +0000 of type Date instead.", underlyingError: nil)): file /BuildRoot/Library/Caches/com.apple.xbs/Sources/swiftlang_Fall20
18/swiftlang_Fall2018-1000.11.42/src/swift/stdlib/public/core/ErrorType.swift, line 191
```

Reply:
yeah, I kinda assume people will try it the normal way first and only set the strategy if that didn't work. error seems fine now.